merqube_client_lib/session.py

```diff
@@ -0,0 +1,291 @@
+
+    def close(self) -> None:
+        if self._session is not None:
+            self._session.close()
+            self._session = None
+            self._session_pid = -1
+
+    def __enter__(self):  # type: ignore
+        """
+        requests.Session like context manager behavior
+        :return: the merq session
+        """
+        return self
+
+    def __exit__(self, exc_type: str, exc_val: Any, exc_tb: Any) -> None:
+        self.close()
+
+    def request(
+        self,
+        method: httpm,
+        url: Union[str, bytes],
+        params: dict[str, Any] | None = None,
+        data: Any = None,
+        headers: dict[str, str] | None = None,
+        options: dict[str, str] | None = None,
+        **kwargs: Any,
+    ) -> Response:
+        """
+        Perform an http request with this session
+        request session helpers sess.put, sess.get, etc, call this
+        """
+        assert params is None or options is None, "both params and options cannot be passed"
+
+        headers = dict(headers) if headers is not None else {}
+
+        # Generate a new requestid if this call isnt being made in a chain that already has it.
+        # (eg client calls dataapi, dataapi calls secapi - we dont want the second call to overwrite the original)
+        # if this isnt set by a client making a call to one of our APIs, the API itself will generate one (eg customer call)
+        # order is: 1) explicitly specified 2) set via chain, 3) generate new
+        headers[REQUEST_ID_HEADER] = headers.get(
+            REQUEST_ID_HEADER, os.getenv(MERQ_REQUEST_ID_ENV_VAR, str(uuid.uuid4()))
+        )
+
+        if self.token:
+            headers["Authorization"] = f"{self.token_type} {self.token}"
+
+        url = urljoin(self._prefix_url, cast(str, url))
+
+        options_dict: dict[str, str] = options or {}
+        return self.http_session.request(
+            method=method.value, url=url, params=options_dict or params, data=data, headers=headers, **kwargs
+        )
+
+    def get(self, url: str, **kwargs: Any) -> Response:
+        """
+        requests.Session like http GET method
+        """
+        return self.request(httpm.GET, url, **kwargs)
+
+    def put(self, url: str, **kwargs: Any) -> Response:
+        """
+        requests.Session like http PUT method
+        """
+        return self.request(httpm.PUT, url, **kwargs)
+
+    def patch(self, url: str, **kwargs: Any) -> Response:
+        """
+        requests.Session like http PATCH method
+        """
+        return self.request(httpm.PATCH, url, **kwargs)
+
+    def post(self, url: str, **kwargs: Any) -> Response:
+        """
+        requests.Session like http POST method
+        """
+        return self.request(httpm.POST, url, **kwargs)
+
+    def delete(self, url: str, **kwargs: Any) -> Response:
+        """
+        requests.Session like http DELETE method
+        """
+        return self.request(httpm.DELETE, url, **kwargs)
+
+
+# Public methods
```

Comment:
Maybe just "public" here since it's not just methods below.

Reply:
I will change this in the next cleanup, thanks!